### PR TITLE
add uri

### PR DESCRIPTION
**Pull request by paulofaria**
# URI

The `URI` type represents an URI.

```swift
public struct URI {
    public struct userInfo {
        public var username: String
        public var password: String
    }

    public struct Query {
        public var key: String
        public var value: String?
    }

    public var scheme: String?
    public var userInfo: userInfo?
    public var host: String?
    public var port: Int?
    public var path: String?
    public var query: [Query]
    public var fragment: String?
}
```

## Motivation

Having `URI` as a `struct` allows it to conform to protocols in extensions.

## References

- [RFC 3986](https://tools.ietf.org/html/rfc3986)

### CODE REVIEW

File: Docs/uri.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+The `URI` type represents an URI.
+
+```swift
+public struct URI {
+    public var scheme: String?
+    public var userInfo: (username: String, password: String)?
+    public var host: String?
+    public var port: Int?
+    public var path: String?
+    public var query: [(key: String, value: String?)]
```
Comment (LoganWright) [inline Docs/uri.md]: What's the situation where `value` will be `nil` if `key` has value? Would it be something like `key=` w/o a value?

Comment (tannernelson) [inline Docs/uri.md]: `?foo&bar=&`

Foo is `nil` and bar is `""`

File: Docs/uri.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# URI
+
+The `URI` type represents an URI.
+
+```swift
+public struct URI {
+    public struct userInfo {
```
Comment (LoganWright) [inline Docs/uri.md]: I think should be `U`

Comment (paulofaria) [inline Docs/uri.md]: you're right!

Comment (paulofaria) [inline Docs/uri.md]: nice catch!

**Pull request merged**